Exclude group backend from showing groups in users page (group manager filter approach)

PVince81 opened the pull request:
To be able to exclude backends for specific scopes.

WIP for https://github.com/owncloud/customgroups/issues/6

Commit:
Group manager search by scope

To be able to exclude backends for specific scopes.

PVince81 commented:
Here we go... Now this works as we want it.

The only concern: what happens to caching since we call with the scope more often.

We could also cache by scope, but there is a risk of using a lot of memory.

Or debug all code paths to find out whether caching is even needed for these scope-specific cases.

PVince81 commented:
the customgroups PR, for testing: https://github.com/owncloud/customgroups/pull/30

PVince81 commented:
@jvillafanez I found a way to keep the cache: eca0d6c12b646d8297e75b1be88778b77bd192ed

Bascially always cache all the user's memberships but filter some out if a scope was given. And this all in the group manager. What do you think ?

It's a kind of post-filter but still within that method.
In the future when we rewire it to the central group table, the `$excludedBackendsForScope` can be plugged directly into the SQL query if needed.

jvillafanez commented on lib/private/Group/Group.php:
```diff
@@ -297,4 +297,14 @@ private function getVerifiedUsers($userIds) {
+	 * @since 10.0.0
+	 */
+	public function getBackend() {
+		return current($this->backends);
```
This needs some explanation. I'd expect the `getBackend` returns the same backend, which might not be the case with this code.

PVince81 replied:
It is, but `$this->backends` is an array because apparently the `Group` class can wrap multiple groups with the same name but from different backends. In practice this doesn't happen because deduplication of ids kicks in earlier.

PVince81 replied:
I'll add a comment

jvillafanez commented on lib/private/Group/Manager.php:
```diff
@@ -239,39 +243,63 @@ public function search($search, $limit = null, $offset = null) {
-			return $this->cachedUserGroups[$uid];
+	public function getUserIdGroups($uid, $scope = null) {
+		// gather the list of backends that opt out of the given scope, if given
+		$excludedBackendsForScope = [];
```
Move this list to a function, and provide caching if needed?

PVince81 replied:
Since the number of backends is low I don't think we need to cache this.

A separate function sounds good.

jvillafanez commented on lib/private/Group/Manager.php:
```diff
@@ -239,39 +243,63 @@ public function search($search, $limit = null, $offset = null) {
 					}
 				}
 			}
+			if ($scope === null) {
```
I think we can try a "cache everything, filter later" approach. We can cache all the groups regardless of the scope, and return the filtered version.

PVince81 replied:
I forgot to remove this `if`. Yes, we should always cache and the post-filter takes care of filtering.

PVince81 commented on lib/private/Group/Manager.php:
```diff
@@ -218,9 +218,13 @@ public function createGroup($gid) {
+	public function search($search, $limit = null, $offset = null, $scope = null) {
 		$groups = [];
 		foreach ($this->backends as $backend) {
+			if ($scope !== null && !$backend->isVisibleForScope($scope)) {
```
@jvillafanez we talked about whitelisting. I realized that this code would not do that. Passing `null` would not skip any backends.

I'll change this to pass `null` to `isVisibleForScope` so backends can also exclude themselves in a generalized way.

PVince81 commented:
Adjusted as follows:

- remove null check for scope and always call `isVisibleForScope()` even with null, for whitelisting
- remove null check in `getUserGroups` to always cache all groups
- moved `excludedBackendsForScope` logic to two functions, one to get them and one to filter
- added unit tests for `search()` and `getUserGroups()` which both have the new `$scope` argument

@jvillafanez

PVince81 commented:
Uh oh, breakage ahead:
```
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #1
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #2
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #3
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #4
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #5
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #6
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #7
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #8
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #9
    Test\Repair\RepairUnmergedSharesTest::testMergeGroupShares.testMergeGroupShares with data set #13
    Test\Share20\ManagerTest.testUserCreateChecksShareWithGroupMembersOnlyDifferentGroups
    Test\Share20\ManagerTest.testUserCreateChecksShareWithGroupMembersOnlySharedGroup
```

jvillafanez commented on lib/public/GroupInterface.php:
```diff
@@ -116,4 +116,14 @@ public function groupExists($gid);
+	/**
+	 * Returns whether the groups are visible for a given scope.
+	 *
+	 * @param string $scope scope string
```
this needs adjustment if `null` can also be used as parameter.

jvillafanez commented on lib/private/Group/Group.php:
```diff
@@ -297,4 +297,16 @@ private function getVerifiedUsers($userIds) {
+	public function getBackend() {
+		// multiple backends can exist for the same group name,
+		// but in practice there is only a single one, so return that one
+		return current($this->backends);
```
I'd rather use `$this->backends[0]`. This should lead to a more predictable behaviour and troubleshooting.

jvillafanez replied:
Any comment here?

PVince81 replied:
argh sorry, missed that. Will change it to what you said

PVince81 replied:
done

PVince81 commented:
Fixed the tests. It was cases where the changed functions were mocked and the mock required the extra `null` argument.

I also fixed PHPDoc to say `string|null` for scope.

@jvillafanez check it out: https://github.com/owncloud/core/pull/27287/commits/93b441e721e2fe4c3bd31162944774b9121891c5

PVince81 commented:
squashed and rebased

PVince81 commented:
> @since or @deprecated tag is needed in PHPDoc for OCP\GroupInterface::isVisibleForScope

okay...

PVince81 commented:
Fixed, let's hope Jenkins passes now.

PVince81 commented:
> PHP Fatal error:  Class Mock_GroupInterface_c1d10c6d contains 1 abstract method and must therefore be declared abstract or implement the remaining methods (OCP\GroupInterface::isVisibleForScope) in /var/lib/jenkins/workspace/owncloud-core_core_PR-27287-FHM63TFCIQTKLNHGKBR4MMPJJRYQ5SGYYCZTAUV4BNGWHWZILZGQ/lib/composer/phpunit/phpunit-mock-objects/src/Framework/MockObject/Generator.php(345) : eval()'d code on line 1
10:20:39 Makefile:177: recipe for target 'test-php' failed
10:20:39 make: *** [test-php] Error 255

Dammit! I thought I solve all this already

PVince81 commented:
oh no... seems they got lost in the rebase 😢

PVince81 commented:
I managed to recover the lost commits/branch by looking at past Jenkins builds.

PVince81 commented:
I've rechecked the code diff and retested with https://github.com/owncloud/customgroups/pull/30, it still works fine.

@jvillafanez one last quick review ?

jvillafanez commented:
👍

PVince81 commented:
Jenkins env fail... rebasing again

mrow4a commented:
@PVince81 do you maybe remember what was the reason to introduce https://github.com/owncloud/core/commit/963e660ea10ab14736ab1568ffee1a8957fa2585#diff-a9fd88b3ec14f3556c28679d1ff0e680R307 ?

PVince81 commented:
@mrow4a I don't know why the IGroup instance is wrapping around multiple group backends when in reality there is only once. The diff you linked to is to make it possible to get the backend from here https://github.com/owncloud/core/commit/963e660ea10ab14736ab1568ffee1a8957fa2585#diff-93cb10073d420e1ba441767b3a1e6de7R284

mrow4a commented:
Makes sense, thanks!
